vortex-scalar/src/primitive.rs

```diff
@@ -270,3 +272,70 @@ impl From<usize> for Scalar {
         Scalar::primitive(value as u64, Nullability::NonNullable)
     }
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BinaryNumericOperator {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    // Missing from arrow-rs:
+    // Min,
+    // Max,
+    // Pow,
+}
+
+impl PrimitiveScalar<'_> {
+    /// Apply the (checked) operator to self and other using SQL-style null semantics.
+    ///
+    /// If the operation overflows, Ok(None) is returned.
+    ///
+    /// If the types are incompatible (ignoring nullability), an error is returned.
+    ///
+    /// If either value is null, the result is null.
+    pub fn checked_numeric_operator(
+        self,
+        other: PrimitiveScalar<'_>,
+        op: BinaryNumericOperator,
+    ) -> VortexResult<Option<Scalar>> {
+        if !self.dtype().eq_ignore_nullability(other.dtype()) {
+            vortex_bail!("types must match: {} {}", self.dtype(), other.dtype());
+        }
+
+        let nullability = self.dtype().nullability();
```

Comment:
I also don't think this should be `let nullability = self.dtype.is_nullable() || other.dtype.is_nullable()`

Reply:
Done.